### PR TITLE
TF: XLA beam search + most generation-compatible models are now also XLA-generate-compatible

### PR DESCRIPTION
**Pull request by gante**
# What does this PR do?

The much-awaited PR -- beam search is now XLA compatible. GPT2 is the only model with XLA beam search tests, more models will follow in subsequent PRs 🎊  Preliminary tests on my machine shows that XLA beam search on GPU is ~26x faster (greedy search and sample are ~30x faster).

Slow tests have been run for the usual generate models (gpt2, t5, rag, speech_to_text, encoder_decoder, vision_encoder_decoder, bart).

EDIT: I've also generalized a few functions, and now ALL models that are compatible with generate are also compatible with XLA generate (with a few exceptions, when the models have no past cache support)

__________________________________

A hard-earned lesson which is kinda obvious in hindsight: `if` branches can make the XLA compiler confused about variable shapes, tagging their shape as `<unknown>`, which in turn causes all sorts of exceptions. Out of curiosity, I tried replacing the `if` by `tf.cond`, but the `<unknown>` shape persisted (because the tensor could indeed have a different shape at tracing time, depending on the branch taken)

### CODE REVIEW

File: src/transformers/generation_tf_utils.py [hunk 1]
```diff
@@ -2484,9 +2484,6 @@ def beam_search(
 
         def flatten_beam_dim(tensor, batch_axis=0):
             """Flattens the first two dimensions of a non-scalar array."""
-            # ignore scalars (e.g. cache index)
-            if tf.rank(tensor) == 0:
-                return tensor
```
Comment (gante) [inline src/transformers/generation_tf_utils.py:2489]: This pattern was inherited from FLAX. Contrarily to FLAX, it seems, this `if` throws the XLA compiler of balance, causing `tensor` to have an unknown shape. All problems I was seeing before were a downstream consequence of that unknown shape.

**Comment by patrickvonplaten** [comment]
Very cool! Can we try it out for at least on Encoder-Decoder architecture as well (just to know that this code holds true here)?

**Comment by gante** [comment]
@patrickvonplaten @Rocketknight1 now with encoder-decoder tests, and ready for review -- I was working on it on a separate branch, so I've merged it into this one. Now, this PR standardizes the XLA model kwargs preparation, and most models can use the XLA functionality. Some models were incompatible for different reasons, so there is a new flag to gate XLA generation (and the flag is set in the problematic architectures).

Finally, I'm also considering adding a general test like `test_xla_generate_fast`, but with `@slow`, beam search, and >100 tokens. It will probably break for a few models (like T5), but at least we would be able to automatically track which models are reliable for XLA beam search -- WDYT?

File: src/transformers/generation_tf_utils.py [hunk 2]
```diff
@@ -1807,12 +1815,135 @@ def _update_model_kwargs_for_generation(
         return model_kwargs
 
     def _update_model_kwargs_for_xla_generation(
```
Comment (patrickvonplaten) [inline src/transformers/generation_tf_utils.py:1817]: Ok for me!

File: src/transformers/models/ctrl/modeling_tf_ctrl.py [hunk 3]
```diff
@@ -571,6 +571,8 @@ class TFCTRLLMHead(tf.keras.layers.Layer):
     def __init__(self, config, input_embeddings, **kwargs):
         super().__init__(**kwargs)
         self.vocab_size = config.vocab_size
+        # TODO (Joao): investigate why CTRL has numerical issues in XLA generate
+        self.supports_xla_generation = False
```
Comment (patrickvonplaten) [inline src/transformers/models/ctrl/modeling_tf_ctrl.py:575]: CTRL has close to 0 usage in TF so not worth checking here IMO

File: src/transformers/models/gptj/modeling_tf_gptj.py [hunk 4]
```diff
@@ -722,6 +722,8 @@ def __init__(self, config, *inputs, **kwargs):
         self.lm_head = tf.keras.layers.Dense(
             config.vocab_size, kernel_initializer=get_initializer(config.initializer_range), name="lm_head"
         )
+        # TODO (Joao): investigate why GPTJ has numerical issues in XLA generate
+        self.supports_xla_generation = False
```
Comment (patrickvonplaten) [inline src/transformers/models/gptj/modeling_tf_gptj.py:726]: Let's maybe open a "Good second issue" after this PR for GPT-J

File: src/transformers/models/openai/modeling_tf_openai.py [hunk 5]
```diff
@@ -556,6 +556,8 @@ class TFOpenAIGPTLMHeadModel(TFOpenAIGPTPreTrainedModel, TFCausalLanguageModelin
     def __init__(self, config, *inputs, **kwargs):
         super().__init__(config, *inputs, **kwargs)
         self.transformer = TFOpenAIGPTMainLayer(config, name="transformer")
+        # OpenAIGPT does not have past caching features
+        self.supports_xla_generation = False
```
Comment (patrickvonplaten) [inline src/transformers/models/openai/modeling_tf_openai.py:560]: Also not important IMO

File: tests/models/t5/test_modeling_tf_t5.py [hunk 6]
```diff
@@ -594,6 +573,43 @@ def test_sample_generate(self):
...
+            model.config.prefix + "Today is a beautiful day.",
+            model.config.prefix + "I have four cats, three dogs, two birds, and a horse.",
+        ]
+        input_ids = tokenizer(sentences, return_tensors="tf", padding=True).input_ids
+
+        # xla_generate = tf.function(model.generate, jit_compile=True)
+        xla_generate = tf.function(model.generate)
+
+        # TODO (joao): there is something not quite right with XLA T5 -- as we increase `max_length` the two outputs
+        # drift appart, where the XLA version clearly degrades its quality. XLA-related variables look fine (they are
```
Comment (patrickvonplaten) [inline tests/models/t5/test_modeling_tf_t5.py:597]: Let's also open an issue for this one after having merged the PR - my intuition would be that there is a problem with the relative position embeddings if you don't have the same problem in Bart. Relative position embeddings are quite tricky and could easily have been messed up by XLA

Comment (patrickvonplaten) [inline tests/models/t5/test_modeling_tf_t5.py:597]: Let's also open an issue for this one after having merged the PR - my intuition would be that there is a problem with the relative position embeddings if you don't have the same problem in Bart. Relative position embeddings are quite tricky and could easily have been messed up by XLA

File: tests/test_modeling_tf_common.py [hunk 7]
```diff
@@ -1600,6 +1600,51 @@ def test_dataset_conversion(self):
                 model.compile(optimizer="sgd", run_eagerly=True)
                 model.train_on_batch(test_batch, test_batch_labels)
 
+    def test_xla_generate_fast(self):
```
Comment (patrickvonplaten) [inline tests/test_modeling_tf_common.py:1603]: Nice! Do you know how much time this adds to the Circle CI?

File: src/transformers/generation_tf_utils.py [hunk 8]
```diff
@@ -346,6 +347,7 @@ class TFGenerationMixin:
     """
 
     seed_generator = tf.random.Generator.from_non_deterministic_state()
+    supports_xla_generation = True
```
Comment (patrickvonplaten) [inline src/transformers/generation_tf_utils.py:350]: Is a nice error thrown if a user tries to compile a non-XLA compatible model?

Comment (patrickvonplaten) [reply src/transformers/generation_tf_utils.py:350]: Ah yeah just below - very nice!

**Review by patrickvonplaten — APPROVED** [review]
Cool! Let's merge it :-)

Can open issues right after for:
- T5 XLA degrading perf (think this is a subtle bug - if BART doesn't have this it's either numerical issues which would be a bit weird since T5 has been trained on TPU or maybe the relative position embeddings)
- open a bug for GPT-J (important model)

**Comment by gante** [comment]
Note: if this PR gets merged as it is, I will open an issue to track issues regarding XLA generation (relevant models failing fast tests, as well as models failing the slow tests)

File: src/transformers/generation_tf_utils.py [hunk 9]
```diff
@@ -1807,12 +1815,135 @@ def _update_model_kwargs_for_generation(
...
+                slice_start_base = tf.constant([0, 0, 1, 0])
+                new_past = ()
+                for past_layer in past:
+                    new_past_layer = list(past_layer)
+                    for i in range(len(new_past_layer[:2])):
+                        update_slice = past_layer[i][:, :, -1:]
+                        # Write the last slice to the first open location in the padded past array
+                        # and then truncate the last slice off the array
+                        new_past_layer[i] = dynamic_update_slice(
+                            past_layer[i][:, :, :-1], update_slice, slice_start_base * new_past_index
```
Comment (Rocketknight1) [inline src/transformers/generation_tf_utils.py:1900]: Should this be the suggestion below? It feels like you're taking a slice of `past_layer[i]` and then updating that slice (which will not necessarily update the original). Or does that just work out because it's happening in XLA?
```suggestion
                            past_layer[i], update_slice, slice_start_base * new_past_index
```

File: src/transformers/generation_tf_utils.py [hunk 10]
```diff
@@ -1807,12 +1815,135 @@ def _update_model_kwargs_for_generation(
...
+                    new_past += (tuple(new_past_layer),)
+            else:
+                slice_start_base = tf.constant([0, 0, 0, 1, 0])
+                new_past = [None for _ in range(len(past))]
+                for i in range(len(past)):
+                    update_slice = past[i][:, :, :, -1:]
+                    # Write the last slice to the first open location in the padded past array
+                    # and then truncate the last slice off the array
+                    new_past[i] = dynamic_update_slice(
+                        past[i][:, :, :, :-1], update_slice, slice_start_base * new_past_index
```
Comment (Rocketknight1) [inline src/transformers/generation_tf_utils.py:1911]: Same here - it looks like a copied slice is being updated and that possibly only works because of XLA shenanigans!
```suggestion
                        past[i], update_slice, slice_start_base * new_past_index
```

**Review by Rocketknight1 — APPROVED** [review]
Overall this looks good, although I can't claim to understand the beam search code too well! I have a couple of questions, but if it's passing tests then I assume the problem is just my understanding rather than an actual bug.

File: src/transformers/generation_tf_utils.py (hunk 9 above)
Comment (gante) [reply src/transformers/generation_tf_utils.py:1900]: I will doublecheck that 👍

Comment (gante) [reply src/transformers/generation_tf_utils.py:1900]: If we run

```python
import tensorflow as tf
from tensorflow.compiler.tf2xla.python.xla import dynamic_update_slice

a = tf.range(10)
print("a (before updates) ", a, id(a))

b = dynamic_update_slice(a[:-1], tf.constant([100]), tf.constant([5]))
print("a (after updates)  ", a, id(a))
print("b ", b, id(b))

# this shouldn't work, but it does
c = dynamic_update_slice(a[:-1], tf.constant([100]), tf.constant([20]))
print("c ", c, id(c))
```

we can see that:
1. the original tensor is not touched (praise be 🙏)
2. the output tensor is a new tensor
3. the `[:-1]` slices the input tensor, and the output has the shape of the sliced input tensor
4. indices out of bound are clipped (woot?)

TL;DR it is a safe operation 🙌

File: src/transformers/models/bart/modeling_tf_bart.py [hunk 11]
```diff
@@ -1434,69 +1433,6 @@ def prepare_inputs_for_generation(
...
-            # prepare the padding tensor for `tf.pad`.
-            # `shape=(4, 2)` because each tensor element in `past` has `rank=4`.
-            # `indices=[[2, 1]]` means the time dimension (dim 2) needs **right**-padding (`1` means padding afterward).
-            padding_values = tf.scatter_nd(indices=[[2, 1]], updates=[num_padding_values], shape=(4, 2))
-
-            new_past = ()
-            for past_layer in past:
-                new_past_layer = list(past_layer)
-                for i in range(len(new_past_layer[:2])):
-                    new_past_layer[i] = tf.pad(past_layer[i], padding_values)
```
Comment (aashiqmuhamed) [inline src/transformers/models/bart/modeling_tf_bart.py:1458]: @gante I was wondering why you decided to go for right padding as opposed to say left padding which could be simpler (no special treatment for relative positional embeddings, no dynamic_update_slicing required).

Comment (gante) [reply src/transformers/models/bart/modeling_tf_bart.py:1458]: Hey @aashiqmuhamed! 

To be candid, we did not even consider other forms to pre-populate the fixed-shape tensors. Maybe left-padding would lead to faster code, as it would imply a concatenation on the right and a cropping on the left (perhaps faster than a scatter operation). 

I'm sure there are many optimization opportunities in the TF XLA generate codebase -- for instance, beam search relies on many expensive reshapes, which is not necessary. 

(`dynamic_update_slicing` is simply syntactic sugar, it can be replaced by a more verbose scatter operation)
